keps/sig-scheduling/20190311-resource_bin_packing_priority_function.md

```diff
@@ -0,0 +1,257 @@
+
+## Motivation
+
+While running workloads on kubernetes which use accelerator devices the default scheduler spreads pods across nodes resulting in fragmentation of extended resources. This fragmentation prevents scheduling pods with larger device resource requirements, and they remain in the pending state.
+
+### Goals
+
+- Schedule Jobs using BestFit Policy using RequestedToCapacityRatio Priority Function for extended Resources
+- Reduce Fragmentation of scarce resources on the Cluster
+
+### Non-Goals
+
+-
+
+## Proposal
+
+The plan is to modify  `requested_to_capacity_ratio` priority function to support `resource_bin_packing` of extended resources. Add another argument `resources` of type `map[v1.ResourceName]int64{}` .This would allow users who want to bin pack a resource to use the function by setting the argument resources which would require them to specify weights for bin packing. For example
+
+```yaml
+"priorities": [
+   ...{
+      "name":"RequestedToCapacityRatioPriority",
+      "weight":5,
+      "argument":{
+         "requestedToCapacityRatioArguments":{
+            "shape":[
+               {
+                  "utilization":100,
+                  "score":10
+               },
+               {
+                  "utilization":0,
+                  "score":0
+               }
+            ],
+            "resources":[
+               {
+                  "resource":"intel.com/foo",
+                  "weight":5
+               },
+               {
+                  "resource":"intel.com/bar",
+                  "weight":2
+               },
+               {
+                  "resource":"cpu",
+                  "weight":1
+               }
+            ]
+         }
+      }
+   }
+]
+```
+
+The argument `resources` is optional and by default will be set to the value
+
+```go
+resources := map[v1.ResourceName]int64{
+  v1.ResourceMemory : 1,
+  v1.ResourceCPU    : 1,
+}
+```
+
+### Argument Input Scenarios
+
+1. If the resource does not exist the score for that resources defaults to 0
+2. As we are storing resources as a map there can be a single weight for every resource type.
+3. Negative weights are not allowed, this would be filtered inside the function and an appropriate error would be thrown.
+
+
+### Design Details
+
+The node score would be calculated as (podRequest + requested) / allocatable. The weights would be used to calculate the resulting node score in the following way.
+
+```go
+resources := make(map[v1.ResourceName]weight)
+nodeScore := 0
+weightSum := 0
+for resource, weight := range resources {
+  nodeScore += resourceScoringFunction((podRequest[resource]+ requested[resource]), allocatable[resource]) * weight
+  weightSum += weight
+}
+nodeScore = (nodeScore / weightSum)* 10
```

Comment:
where this `*10` comes from?

Reply:
that's from the old function, it shouldn't be there